## The Phase clinical trial dataset for new oral insulin called Auralin.

The Auralin and Novodra are not real insulin products. This clinical trial data was fabricated. When assessing this data, the issues that you'll detect (and later clean) are meant to simulate real-world data quality and tidiness issues.

## Gather

In [1]:
import pandas as pd
import numpy as np

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

## Visual Assessment

##### `patients` columns:

- `patient_id`: the unique identifier for each patient in the Master Patient Index (i.e. patient database) of the pharmaceutical company that is producing Auralin.
- `assigned_sex`: the assigned sex of each patient at birth (male or female).
- `given_name`: the given name (i.e. first name) of each patient.
- `surname`: the surname (i.e. last name) of each patient.
- `address`: the main address for each patient.
- `city`: the corresponding city for the main address of each patient.
- `state`: the corresponding state for the main address of each patient.
- `zip_code`: the corresponding zip code for the main address of each patient.
- `country`: the corresponding country for the main address of each patient (all United states for this clinical trial).
- `contact`: phone number and email information for each patient.
- `birthdate`: the date of birth of each patient (month/day/year). The inclusion criteria for this clinical trial is age >= 18 - (there is no maximum age because diabetes is a growing problem among the elderly population).
- `weight`: the weight of each patient in pounds (lbs).
- `height`: the height of each patient in inches (in).
- `bmi`: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m2 where kg is a person's weight in kilograms and m2 is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. The inclusion criteria for this clinical trial is 16 >= BMI >= 38.

In [4]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

## Visual Assessment

##### `treatments` columns:

- `given_name`: the given name of each patient in the Master Patient Index that took part in the clinical trial.
- `surname`: the surname of each patient in the Master Patient Index that took part in the clinical trial.
- `auralin`: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) and the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the international unit of measurement and the standard measurement for insulin.
- `novodra`: same as above, except for patients that continued treatment with Novodra.
- `hba1c_start`: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The HbA1c test measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- `hba1c_end`: the patient's HbA1c level at the end of the last week of treatment.
- `hba1c_change`: the change in the patient's HbA1c level from the start of treatment to the end, i.e., hba1c_start - hba1c_end. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

In [5]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

## Visual Assessment

##### `adverse_reactions` columns:

- `given_name`: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra).
- `surname`: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra).
- `adverse_reaction`: the adverse reaction reported by the patient.


In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 37.4+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 11.0+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 472.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex given_name     surname                 address  \
354         355       female     Vivian       House      4932 Goldleaf Lane   
206         207       female       Leah        Wong      3548 Hilltop Drive   
211         212       female     Martha  Afanasyeva    3613 Lodgeville Road   
393         394         male     Steven         Roy      2816 Ashford Drive   
493         494       female        Fen        Chin  1826 Poplar Chase Lane   

           city state  zip_code        country  \
354      Newark    NJ    7102.0  United States   
206    Canadian    TX   79014.0  United States   
211  Saint Paul    MN   55114.0  United States   
393      Reston    VA   22070.0  United States   
493       Boise    ID   83702.0  United States   

                                       contact  birthdate  weight  height  \
354        201-586-2848VivianRHouse@dayrep.com  8/13/1936   130.2      62   
206            806-323-2210LeahWong@einrot.com  6/11/1934   129.8      62   
211  MarthaAfanasyeva@dayrep.com1 612 228 4170  10/3/1997   151.2      73   
393         StevenBRoy@rhyta.com1 703 716 8652  9/19/1975   162.4      72   
493         FenChin@gustr.com+1 (208) 388-1065  3/18/1997   195.1      68   

      bmi  
354  23.8  
206  23.7  
211  19.9  
393  22.0  
493  29.7

In [15]:
patients.surname.value_counts()

Doe         6
Jakobsen    3
Taylor      3
Hueber      2
Silva       2
           ..
Teng        1
Nnonso      1
Hsu         1
Markus      1
Jindrová    1
Name: surname, Length: 466, dtype: int64

In [16]:
patients.address.value_counts()

123 Main Street             6
648 Old Dear Lane           2
2778 North Avenue           2
2476 Fulton Street          2
4852 Rose Avenue            1
                           ..
2324 Benson Street          1
2055 Emeral Dreams Drive    1
4535 Prospect Street        1
1463 Martha Ellen Drive     1
1116 Dog Hill Lane          1
Name: address, Length: 483, dtype: int64

In [17]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [19]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [20]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight, patient_id = '210'

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- `Adverse reaction` column should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

-- --

## Clean

In [23]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data
- `treatments`: Missing records (280 instead of 350)
- `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)

Note: these lines for organization, clarity and optional cuz these lines was written above in Assess paragraph.


##### Define

- Import the cut treatments into a DataFrame and concatenate it with the original treatments DataFrame.

- Recalculate the hba1c_change column: hba1c_start minus hba1c_end.

##### Code

In [24]:
# treatments: Missing records (280 instead of 350)
treatments_cut = pd.read_csv('treatments_cut.csv')

# Clear the existing index and reset it in the result by setting the ignore_index option to True.
treatments_clean = pd.concat([treatments_clean, treatments_cut],
                             ignore_index=True)

##### Test

In [25]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [26]:
treatments_clean.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

##### Code

In [27]:
# treatments: Missing HbA1c changes and Inaccurate HbA1c changes
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - treatments_clean.hba1c_end)

##### Test

In [28]:
treatments_clean.hba1c_change

0      0.43
1      0.47
2      0.43
3      0.35
4      0.32
       ... 
345    0.34
346    0.45
347    0.30
348    0.47
349    0.46
Name: hba1c_change, Length: 350, dtype: float64

### Tidiness

#### Contact column in `patients` table should be split into phone number and email

##### Define
Extract the *phone number* and *email* variables from the *contact* column using regular expressions and pandas' `str.extract` method. Drop the *contact* column when done.

In [29]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

##### Test

In [30]:
# confirm contact column is gone
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [31]:
patients_clean.phone_number.sample(10)

15          276-225-1955
339         952 431 5166
239         228-378-1355
272    +1 (937) 518-7238
381    +1 (870) 457-5086
461         646-289-4177
429         218-773-9682
35          718-501-0503
157         920-849-0384
394         903-939-1025
Name: phone_number, dtype: object

In [32]:
# confirm that no emails start with integer
patients_clean.email.value_counts().head()

johndoe@email.com            6
PatrickGersten@rhyta.com     2
SandraCTaylor@dayrep.com     2
JakobCJakobsen@einrot.com    2
EllenRLuman@einrot.com       1
Name: email, dtype: int64

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
Melt the *auralin* and *novodra* columns to a *treatment* and a *dose* column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain *start_dose* and *end_dose* columns. Drop the intermediate *dose* column.

##### Code

In [33]:
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split('-', 1).str

treatments_clean = treatments_clean.drop('dose', axis=1)

C:\Users\Esam\miniconda3\lib\site-packages\ipykernel_launcher.py:4: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  after removing the cwd from sys.path.


##### Test

In [34]:
treatments_clean.sample(5)

given_name    surname  hba1c_start  hba1c_end  hba1c_change treatment  \
667       sofia  hermansen         8.90       8.57          0.33   novodra   
614  chukwumoge  ogochukwu         7.95       7.56          0.39   novodra   
296      mijael     guerra         7.85       7.46          0.39   auralin   
15         enco     žibrik         7.78       7.34          0.44   auralin   
398         zak      kelly         7.66       7.26          0.40   novodra   

    dose_start dose_end  
667       34u       34u  
614       41u       39u  
296       37u       47u  
15        55u       68u  
398       38u       38u

#### `Adverse reaction` column should be part of the `treatments` table

##### Define
Merge the *adverse_reaction* column to the `treatments` table, joining on *given name* and *surname*.

##### Code

In [35]:
adverse_reactions_clean.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [36]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on=['given_name', 'surname'], how='left') 

##### Test

In [37]:
treatments_clean

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       veronika    jindrová         7.63       7.20          0.43   auralin   
1           skye  gormanston         7.97       7.62          0.35   auralin   
2         sophia      haugen         7.65       7.27          0.38   auralin   
3          eddie      archer         7.89       7.55          0.34   auralin   
4           asia     woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
345  christopher    woodward         7.51       7.06          0.45   novodra   
346        maret    sultygov         7.67       7.30          0.37   novodra   
347        lixue       hsueh         9.21       8.80          0.41   novodra   
348        jakob    jakobsen         7.96       7.51          0.45   novodra   
349        berta  napolitani         7.68       7.21          0.47   novodra   

    dose_start dose_end           adverse_reaction  
0         41u       48u                        NaN  
1         33u       36u                        NaN  
2         37u       42u                        NaN  
3         31u       38u                        NaN  
4         30u       36u                        NaN  
..         ...      ...                        ...  
345       55u       51u                     nausea  
346       26u       23u                        NaN  
347       22u       23u  injection site discomfort  
348       28u       26u               hypoglycemia  
349       42u       44u  injection site discomfort  

[350 rows x 9 columns]

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables and Lowercase given names and surnames

##### Define
- Adverse reactions table is no longer needed so ignore that part. Isolate the patient ID and names in the patients table,
- Then convert these names to lower case to join with treatments. 
- Then drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore).

##### Code

In [38]:
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

C:\Users\Esam\miniconda3\lib\site-packages\pandas\core\generic.py:5170: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


##### Test

In [39]:
# Confirm the merge was executed correctly
treatments_clean

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0           7.63       7.20          0.43   auralin       41u       48u   
1           7.97       7.62          0.35   auralin       33u       36u   
2           7.65       7.27          0.38   auralin       37u       42u   
3           7.89       7.55          0.34   auralin       31u       38u   
4           7.76       7.37          0.39   auralin       30u       36u   
..           ...        ...           ...       ...        ...      ...   
344         7.51       7.06          0.45   novodra       55u       51u   
345         7.67       7.30          0.37   novodra       26u       23u   
346         9.21       8.80          0.41   novodra       22u       23u   
347         7.96       7.51          0.45   novodra       28u       26u   
348         7.68       7.21          0.47   novodra       42u       44u   

              adverse_reaction  patient_id  
0                          NaN         225  
1                          NaN         242  
2                          NaN         345  
3                          NaN         276  
4                          NaN          15  
..                         ...         ...  
344                     nausea         153  
345                        NaN         420  
346  injection site discomfort         336  
347               hypoglycemia          25  
348  injection site discomfort         477  

[349 rows x 8 columns]

In [40]:
# Patient ID should be the only duplicate column

all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

-- --

### Quality

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
- Convert the zip code column's data type from a float to a string using `astype`.
- remove the '.0' using string slicing, and pad four digit zip codes with a leading 0.

##### Code

In [41]:
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')

# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

##### Test

In [42]:
patients_clean.zip_code

0      92390
1      61812
2      68467
3      07095
4      36303
       ...  
498    03852
499    86341
500    64110
501    98109
502    68324
Name: zip_code, Length: 503, dtype: object

In [43]:
patients_clean.zip_code.isnull().value_counts()

False    491
True      12
Name: zip_code, dtype: int64

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
Replace height for rows in the `patients` table that have a height of 27 in (there is only one) with 72 in.

In [44]:
patients_clean.height = patients_clean.height.replace(27, 72)

##### Test

In [45]:
patients_clean[patients_clean.height == 27]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [46]:
patients_clean[patients_clean.surname == 'Neudorf']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

   phone_number                email  
4  334-515-7487  TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
Apply a function that converts full state name to state abbreviation for California, New York, Illinois, Florida, and Nebraska.

##### Code

In [47]:
# Mapping from full state name to abbreviation

abbreviations = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

patients_clean['state'] = patients_clean['state'].replace(abbreviations) 

##### Test

In [48]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
MI    13
OK    13
LA    13
NJ    12
VA    11
MS    10
WI    10
AL     9
IN     9
MN     9
TN     9
KY     8
NC     8
WA     8
MO     7
ID     6
KS     6
NV     6
NE     6
CT     5
SC     5
IA     5
CO     4
ND     4
ME     4
RI     4
AR     4
AZ     4
OR     3
DE     3
SD     3
WV     3
MD     3
DC     2
VT     2
MT     2
WY     1
AK     1
NH     1
NM     1
Name: state, dtype: int64

#### Dsvid Gustafsson

##### Define
Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'.

##### Code

In [49]:
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### Test

In [50]:
patients_clean[patients_clean.given_name == 'David']

patient_id assigned_sex given_name     surname               address  \
8             9         male      David  Gustafsson    1790 Nutter Street   
280         281         male      David    Beauvais    98 Freshour Circle   
440         441         male      David   Villadsen  1434 Oak Ridge Drive   

             city state zip_code        country   birthdate  weight  height  \
8     Kansas City    MO    64105  United States    3/6/1937   163.9      66   
280   San Antonio    TX    78258  United States    5/1/1931   128.5      67   
440  St Elizabeth    MO    65559  United States  11/30/1924   137.3      68   

      bmi       phone_number                        email  
8    26.5       816-265-9578  DavidGustafsson@armyspy.com  
280  20.1       210-218-3477     DavidLBeauvais@rhyta.com  
440  20.9  +1 (573) 493-4748  DavidMVilladsen@armyspy.com

In [51]:
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0         7.63       7.20          0.43   auralin       41u       48u   
1         7.97       7.62          0.35   auralin       33u       36u   
2         7.65       7.27          0.38   auralin       37u       42u   
3         7.89       7.55          0.34   auralin       31u       38u   
4         7.76       7.37          0.39   auralin       30u       36u   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
- Convert assigned sex and state to categorical data types.
- Zip code data type was already addressed above.
- Convert birthdate to datetime data type.
- Strip the letter 'u' in start dose and end dose and convert those columns to data type integer.

##### Code

In [52]:
# To category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# To datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# Strip u and to integer
treatments_clean.dose_start = treatments_clean.dose_start.str.strip("u ").astype(int)
treatments_clean.dose_end = treatments_clean.dose_end.str.strip("u ").astype(int)

##### Test 

In [53]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    category      
 7   zip_code      491 non-null    object        
 8   country       491 non-null    object        
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int64         
 12  bmi           503 non-null    float64       
 13  phone_number  491 non-null    object        
 14  email         491 non-null    object        
dtypes: category(2), datetime64[ns](1), float

In [54]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   hba1c_start       349 non-null    float64
 1   hba1c_end         349 non-null    float64
 2   hba1c_change      349 non-null    float64
 3   treatment         349 non-null    object 
 4   dose_start        349 non-null    int32  
 5   dose_end          349 non-null    int32  
 6   adverse_reaction  35 non-null     object 
 7   patient_id        349 non-null    int64  
dtypes: float64(3), int32(2), int64(1), object(2)
memory usage: 19.1+ KB


In [55]:
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment  dose_start  dose_end  \
0         7.63       7.20          0.43   auralin          41        48   
1         7.97       7.62          0.35   auralin          33        36   
2         7.65       7.27          0.38   auralin          37        42   
3         7.89       7.55          0.34   auralin          31        38   
4         7.76       7.37          0.39   auralin          30        36   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

#### Multiple phone number formats

##### Define
- Strip all " ", "-", "(", ")", and "+" and store each number without any formatting.
- Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

In [56]:
# \D is any character that is not numeric degit from 0 to 9
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [57]:
patients_clean.phone_number

0      19517199170
1      12175693204
2      14023636804
3      17326368246
4      13345157487
          ...     
498    12074770579
499    19282844492
500    18162236007
501    13604432060
502    14028484923
Name: phone_number, Length: 503, dtype: object

#### Default John Doe data

##### Deffine
Remove the non-recoverable John Doe records from the patients table.

##### Code

In [58]:
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [59]:
patients_clean.surname.value_counts()

Taylor          3
Jakobsen        3
Grímsdóttir     2
Lund            2
Souza           2
               ..
Baumgaertner    1
Teng            1
Nnonso          1
Hsu             1
Jindrová        1
Name: surname, Length: 465, dtype: int64

In [60]:
# Should be no 123 Main Street records
patients_clean.address.value_counts()

2476 Fulton Street          2
2778 North Avenue           2
648 Old Dear Lane           2
1953 Rhapsody Street        1
230 Lady Bug Drive          1
                           ..
1826 Poplar Chase Lane      1
2324 Benson Street          1
2055 Emeral Dreams Drive    1
4535 Prospect Street        1
1116 Dog Hill Lane          1
Name: address, Length: 482, dtype: int64

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
- Remove the Jakob Jakobsen, Thể Lương, and Sandra Taylor rows from the patients table. These are the nicknames, which happen to also not be in the treatments table (removing the wrong name would create a consistency issue between the patients and treatments table). 
- These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

##### Code

In [61]:
patients_clean.surname.value_counts()

Taylor          3
Jakobsen        3
Grímsdóttir     2
Lund            2
Souza           2
               ..
Baumgaertner    1
Teng            1
Nnonso          1
Hsu             1
Jindrová        1
Name: surname, Length: 465, dtype: int64

In [62]:
# tilde means not: http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing
patients_clean = patients_clean[~((patients_clean.surname.duplicated()) & patients_clean.surname.notnull())]

##### Test

In [63]:
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname            address  \
24          25         male      Jakob  Jakobsen  648 Old Dear Lane   

           city state zip_code        country  birthdate  weight  height  \
24  Port Jervis    NY    12771  United States 1985-08-01   155.8      67   

     bmi phone_number                      email  
24  24.4  18458587707  JakobCJakobsen@einrot.com

In [64]:
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   

    phone_number                     email  
131  13044382648  SandraCTaylor@dayrep.com

In [65]:
patients_clean[patients_clean.surname == 'Lương']

patient_id assigned_sex given_name surname             address     city  \
197         198         male        Thể   Lương  2403 Ripple Street  Gaylord   

    state zip_code        country  birthdate  weight  height   bmi  \
197    MI    49735  United States 1967-11-15   223.1      70  32.0   

    phone_number                    email  
197  19897057690  LuongKhacThe@dayrep.com

##### Test

In [66]:
patients_clean.surname.value_counts()

Sheppard       1
Alanis         1
Synek          1
Tuma           1
Eldarkhanov    1
              ..
Moore          1
Enríquez       1
Gustafsson     1
Magyar         1
Jindrová       1
Name: surname, Length: 465, dtype: int64

#### kgs instead of lbs for Zaitseva weight

##### Define
Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

In [67]:
weight_kg = patients_clean.weight.min()
patient_name = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[patient_name, column_name] = weight_kg * 2.20462

##### Test

In [68]:
# 48.8 shouldn't be the lowest anymore
patients_clean.weight.sort_values()

459    102.1
335    102.7
74     103.2
171    106.5
51     107.1
       ...  
61     244.9
144    244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 465, dtype: float64